File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int maxpr = nums[0];
        int minpr = nums[0];
        int ans = nums[0];
        for(int i=1; i<n; i++){
            if(nums[i] < 0){
                swap(maxpr, minpr);
            }
            maxpr = max(nums[i] , maxpr * nums[i]);
            minpr = min(nums[i], minpr * nums[i]);
            ans = max(ans, maxpr);
        }
        return ans;
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        maxpr = nums[0]
        minpr = nums[0]
        ans = nums[0]
        for i in range(1, len(nums)):
            if nums[i] < 0:
                maxpr, minpr = minpr, maxpr
            maxpr = max(nums[i], nums[i] * maxpr)
            minpr = min(nums[i], nums[i] * minpr)
            ans = max(ans, maxpr)
        return ans
        